# BF520 library design 

This directory contains an analysis of previous BF520 DMS data and natural sequence variation. The goal of this analysis is to use previous deep mutational scanning data and natural sequence variation to cut down on the number of lethal mutations introduced into BF520 variants in barcoded lentivirus deep mutational scanning libraries. Mutations will be selected for this new library that were not highly seleted against in the previous deep mutational scan, or that are present above some frequency in the natural sequence alignments. 

## Organization 

The analysis of mutant effects and design of the new mutant library is contained in the Juptyer notebook [analyze_BF520_prefs.ipynb](analyze_BF520_prefs.ipynb).  

The results are converted into a format used by the top level directory analysis in the notebook [reformat_designed_library_df.ipynb](reformat_designed_library_df.ipynb).  

### Inputs 

There are several required inputs for [analyze_BF520_prefs.ipynb](analyze_BF520_prefs.ipynb):  
- Codon counts of mutations to BF520 pre and post deep mutational scanning selection, as determined by Hugh Haddox and Adam Dingens. See [here](https://elifesciences.org/articles/34420#SA1) for information on how these data were generated. Shortly; mutant virus libraries containing variants of BF520 Env were sequenced pre and post growth selection using [barcoded subamplicon sequencing](https://jbloomlab.github.io/dms_tools2/bcsubamp.html). The FASTQ files containing the raw reads are available in the [Sequence Read Archive (SRA)](http://www.ncbi.nlm.nih.gov/sra), and have BioSample Accession number [SAMN06313000](https://www.ncbi.nlm.nih.gov/sra?term=SAMN06313000) and BioProject number [PRJNA371844](https://www.ncbi.nlm.nih.gov/bioproject/PRJNA371844/). The codon counts files were generated by processing the FASTQ files using [dms2_batch_bcsubamp](https://jbloomlab.github.io/dms_tools2/dms2_batch_bcsubamp.html#dms2-batch-bcsubamp). This pipeline can be viewed and run in [Supplemental Figure 1 of this paper](https://elifesciences.org/download/aHR0cHM6Ly9jZG4uZWxpZmVzY2llbmNlcy5vcmcvYXJ0aWNsZXMvMzQ0MjAvZWxpZmUtMzQ0MjAtc3VwcDEtdjMuemlw/elife-34420-supp1-v3.zip?_hash=6MnzQzLRxpqLVKLJ4NDDQWOwXk5KvpGaMsBd06aYC%2Bo%3D). For this analysis, I have copied the codon counts files from Hugh's computational analysis into [data/counts_from_Hugh/renumberedcounts/](data/counts_from_Hugh/renumberedcounts/).  
- A file for rebnumbering BF520 residues to the [HXB2 reference strain](http://people.biology.ucsd.edu/satish/Science/MePapers/korber.pdf) residue numbers: [BF520_to_HXB.csv](data/BF520_to_HXB.csv). This file was generated by Hugh Haddox, using a manually tweaked version of an alignment created by [mafft](https://mafft.cbrc.jp/alignment/software/). He specifically noted that he manually re-aligned BF520 in the regions between 184-191 (noninclusive bounds in variable loop 1; HXB2 numbering) and 395-412 (noninclusive bounds in variable loop 4; HXB2 numbering).  
- [Filtered web alignments from LANL](https://www.hiv.lanl.gov/content/sequence/NEWALIGN/align.html) of both DNA and amino acid HIV sequences. These alignments are used to determine the natural sequence variation of HIV Env. They were downloaded on November 9th, 2020 from [the curated alignments from LANL](https://www.hiv.lanl.gov/content/sequence/NEWALIGN/align.html) with the following settings:  
    - Alignment type: Filtered web  
    - Organism: HIV-1/SIVcpz  
    - Region: Env  
    - Subtype: M group without recombinants (A-K)  
    - DNA/Protein: DNA for DNA alignment, Protein for amino acid alignment  
    - Year: 2018  
    - Format: FASTA  

### Outputs  

There are several important outputs from [analyze_BF520_prefs.ipynb](analyze_BF520_prefs.ipynb):  

- [IDT_library_df.csv](results/IDT_library_df.csv), which is used as an input file into `TargetedTilingPrimers` to produce mutagenesis primers for making the mutant libraries.  
- [conversion.csv](results/conversion.csv), which is used to convert [IDT_library_df.csv](results/IDT_library_df.csv) into [designed_mutations.csv](results/designed_mutations.csv) using the [reformat_designed_library_df.ipynb](reformat_designed_library_df.ipynb) notebook.  
- [designed_mutations.csv](results/designed_mutations.csv), which is used in the top level directory analysis of selection data.  

## Running the analysis 

This analysis requires a separate conda environment from the top level analysis of the barcoded lentivirus DMS data due to conflicts between packages. Therefore, you must use conda to set up a separate computing environment for this analysis. 

Ensure you have `conda` installed; if not install it via Miniconda as described [here](https://docs.conda.io/projects/conda/en/latest/user-guide/install/#regular-installation).
The environment is specified in [environment.yml](environment.yml), and an environment with pinned dependencies is specified in [pinned_environment.yml](pinned_environment.yml). You can build the environment with the pinned file in order to rerun the analysis with the same conda environment it was originally run with, or you can build the environment with the unpinned file and conda will re-determine the versions of dependencies to use. 

If you have not previously built the conda environment, then build the environment to `./env` with either:

    conda env create -f environment.yml -p ./env
    
    or 
    
    conda env create -f pinned_environment.yml -p ./env

Then activate it with:

    conda activate ./env

If you've previously built the environment into `./env`, just do the activation step.

Once you have built and activated the environment, you can run the analysis in the [Jupyter notebook][] in this directory. 
